core/ia32.asm

```diff
@@ -0,0 +1,346 @@
+    %define reg_ret_64   rax
+    %define reg_ret      reg_ret_64
+%elifidn __CONV__, x64_microsoft
+    %define reg_arg1_16  cx
+    %define reg_arg1_32  ecx
+    %define reg_arg1_64  rcx
+    %define reg_arg1     reg_arg1_64
+    %define reg_arg2_16  dx
+    %define reg_arg2_32  edx
+    %define reg_arg2_64  rdx
+    %define reg_arg2     reg_arg2_64
+    %define reg_ret_16   ax
+    %define reg_ret_32   eax
+    %define reg_ret_64   rax
+    %define reg_ret      reg_ret_64
+%endif
+
+;
+; Helpers
+;
+
+; Macro: function
+; Declares a function. Arguments:
+; - %1  Name of the function
+; - %2  Number of arguments
+;
+%macro function 2
+    global %1
+    %1:
+%ifidn __CONV__, x32_cdecl
+    %if %2 >= 3
+        %error "Unsupported number of arguments"
+    %else
+        %if %2 >= 1
+            mov reg_arg1, [esp + 0x4]
+        %endif
+        %if %2 >= 2
+            mov reg_arg2, [esp + 0x8]
+        %endif
+    %endif
+%endif
+%endmacro
+
+%macro function_get_reg 1
+    function get_%+%1, 0
+    mov reg_ret, %1
+    ret
+%endmacro
+%macro function_set_reg 1
+    function set_%+%1, 1
+    mov %1, reg_arg1
+    ret
+%endmacro
+%macro function_get_segment 1
+    function get_kernel_%+%1, 0
+    mov reg_ret_16, %1
+    ret
+%endmacro
+%macro function_set_segment 1
+    function set_kernel_%+%1, 1
+    mov %1, reg_arg1_16
+    ret
+%endmacro
+
+section .text
+
+struc qword_struct
+    .lo      resd 1
+    .hi      resd 1
+endstruc
+
+struc cpuid_args
+    ._eax    resd 1
+    ._ecx    resd 1
+    ._edx    resd 1
+    ._ebx    resd 1
+endstruc
+
+function __nmi, 0
+    int 2h
+    ret
+
+function __fls, 1
+    xor reg_ret_32, reg_ret_32
+    bsr reg_ret_32, reg_arg1_32
+    ret
+
+function asm_cpuid, 1
+%ifidn __BITS__, 64
+    push rbx
+    mov r8, reg_arg1
+    mov eax, [r8 + cpuid_args._eax]
+    mov ecx, [r8 + cpuid_args._ecx]
+    cpuid
+    mov [r8 + cpuid_args._eax], eax
+    mov [r8 + cpuid_args._ebx], ebx
+    mov [r8 + cpuid_args._ecx], ecx
+    mov [r8 + cpuid_args._edx], edx
+    pop rbx
+    ret
+%else
+    push ebx
+    push esi
+    mov esi, reg_arg1
+    mov eax, [esi + cpuid_args._eax]
+    mov ecx, [esi + cpuid_args._ecx]
+    cpuid
+    mov [esi + cpuid_args._eax], eax
+    mov [esi + cpuid_args._ebx], ebx
+    mov [esi + cpuid_args._ecx], ecx
+    mov [esi + cpuid_args._edx], edx
+    pop esi
+    pop ebx
+    ret
+%endif
+
+function asm_btr, 2
+    lock btr [reg_arg1], reg_arg2
+    ret
+
+function asm_bts, 2
+    lock bts [reg_arg1], reg_arg2
+    ret
+
+function asm_disable_irq, 0
+    cli
+    ret
+
+function asm_enable_irq, 0
+    sti
+    ret
+
+function asm_fxinit, 0
+    finit
+    ret
+
+function asm_fxrstor, 1
+    fxrstor [reg_arg1]
+    ret
+
+function asm_fxsave, 1
+    fxsave [reg_arg1]
+    ret
+
+function asm_rdmsr, 2
+%ifidn __BITS__, 64
+    mov rcx, reg_arg1
+    rdmsr
+    shl rdx, 32
+    or reg_ret, rdx
+    ret
+%else
+    mov ecx, reg_arg1
+    rdmsr
+    mov [reg_arg2 + qword_struct.lo], eax
```

Comment:
The Win32 driver causes a BSOD as soon as it's loaded, and this is the culprit: Here, `reg_arg2 == edx`, but EDX was just overwritten by `rdmsr`. So we have to save `reg_arg2` to another register (e.g. EBX, which itself may need to be saved to the stack first) before calling `rdmsr`, and use that register as the base address here.

nit: To keep the code portable (e.g. in case we want to support other calling conventions), we could surround cdecl-specific logic with `%ifidn __CONV__, x32_cdecl` .. `%endif`

Reply:
> we could surround cdecl-specific logic with %ifidn __CONV__, x32_cdecl .. %endif

Or perhaps `%ifidn reg_arg2, edx` .. `%endif`. Anyway, it's not easy to achieve portability across different calling conventions, since we probably have cdecl-specific logic elsewhere in the code. So it's perfectly fine if you just explain in a comment why we must save `reg_arg2`.

BTW, I've noticed that line 234, which is equivalent to `mov ecx, ecx`, isn't optimized out by NASM, even for the Release build, and even if I manually run `nasm` with `-Ox`. This isn't a big issue at all, but do you think NASM should be smarter in recognizing this no-op? Or maybe it doesn't want to be too smart, because we could just want to waste some CPU cycles?

Reply:
> but do you think NASM should be smarter in recognizing this no-op? Or maybe it doesn't want to be too smart, because we could just want to waste some CPU cycles?

I believe that's out-of-scope for NASM: it assembles the source as-is and only optimizes code when the instruction mnemonics are ambiguous and allow different variants. E.g. it will prioritize *imm8* over *imm32* when the constants is in range *[0x00, 0xFF]*. There's more info at:
https://www.nasm.us/doc/nasmdoc2.html#section-2.1.23

As you suggested, the only real solution is writing calling-convention specific implementations of functions that have side-effects on "specific" GPRs (just `asm_rdmsr` and `asm_rdmsr`). There might be more such pseudo-nop's out there, but I don't see them as a big issue either. Fortunately, most functions are simple enough that can be implemented just with the `reg_*` macros.

Reply:
I see, thanks! BTW, with `asm_rdmsr` fixed, the new code seems to work well on Win32, so I think we are very close to getting this PR merged!